File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.py
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        n = len(arr)
        left = [0] * n
        right = [0] * n
        stack = []
        mod = int(1e9 + 7)
        
        # Calculate the left limits
        for i in range(n):
            # Maintain the stack such that it contains indices in increasing order of values
            while stack and arr[i] <= arr[stack[-1]]:
                stack.pop()
            # If stack is empty, it means all previous elements are greater, else find the distance
            left[i] = i if not stack else i - stack[-1] - 1
            stack.append(i)
        
        # Clear the stack for re-use
        stack = []
        
        # Calculate the right limits
        for i in range(n-1, -1, -1):
            # Maintain the stack such that it contains indices in increasing order of values
            while stack and arr[i] < arr[stack[-1]]:
                stack.pop()
            # If stack is empty, it means all next elements are greater, else find the distance
            right[i] = (n - i - 1) if not stack else stack[-1] - i - 1
            stack.append(i)
        
        # Calculate the result
        res = 0
        for i in range(n):
            res += arr[i] * (left[i] + 1) * (right[i] + 1)
            res %= mod
        
        return res

File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.java
import java.util.Stack;

class Solution {
    public int sumSubarrayMins(int[] arr) {
        int n = arr.length;
        int[] left = new int[n];
        int[] right = new int[n];
        Stack<Integer> stack = new Stack<>(); // Stack will store indices
        int res = 0;
        int mod = 1000000007; 
        
        // Find left side limits for each element being the minimum in those possible subarrays
        for (int i = 0; i < n; i++){
            // We need to create an increasing sequence inside the stack
            while (!stack.isEmpty() && arr[stack.peek()] >= arr[i]){
                stack.pop();
            }
            
            left[i] = stack.isEmpty() ? i + 1 : (i - stack.peek());
            stack.push(i);
        }
        
        stack.clear(); // Clear stack for the right side calculation
        
        // Now find the right side limits for each element in a similar way
        for (int i = n - 1; i >= 0; i--){
            while (!stack.isEmpty() && arr[stack.peek()] > arr[i]){
                stack.pop();
            }
            
            right[i] = stack.isEmpty() ? (n - i) : (stack.peek() - i);
            stack.push(i);
        }
        
        // Calculate the result based on left and right limits
        for (int i = 0; i < n; i++){
            res = (res + (int)((long)arr[i] * left[i] * right[i] % mod)) % mod;
        }
        
        return res;
    }
}
